test/statusline.cpp

```diff
@@ -0,0 +1,202 @@
+				REQUIRE(status_message_handler.last_message == "test message...");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+
+			WHEN("the message_lifetime object goes out of scope") {
+				message_lifetime.reset();
+
+				THEN("the message is cleared") {
+					REQUIRE(status_message_handler.last_message == "");
+					REQUIRE_FALSE(status_message_handler.last_string_was_error);
+				}
+			}
+		}
+	}
+
+	GIVEN("a previous show_message_until_finished call") {
+		const auto previous_message_lifetime =
+			status_line.show_message_until_finished("previous message...");
+
+		WHEN("show_message() is called") {
+			status_line.show_message("this is a status");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is a status");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+		}
+
+		WHEN("show_error() is called") {
+			status_line.show_error("this is an error");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is an error");
+				REQUIRE(status_message_handler.last_string_was_error);
+			}
+		}
+
+		WHEN("show_message_until_finished() is called") {
+			auto message_lifetime = status_line.show_message_until_finished("test message...");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "test message...");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+
+			WHEN("the message_lifetime object goes out of scope") {
+				message_lifetime.reset();
+
+				THEN("the previous message is restored") {
+					REQUIRE(status_message_handler.last_message == "previous message...");
+					REQUIRE_FALSE(status_message_handler.last_string_was_error);
+				}
+			}
+		}
+	}
+}
+
+TEST_CASE("show_message_until_finished() will restore old message when lifetime is destructed")
+{
+	StatusTester status_message_handler;
+	StatusLine status_line(status_message_handler);
+
+	GIVEN("multiple message lifetimes") {
+		auto message_lifetime1 = status_line.show_message_until_finished("first message");
+		REQUIRE(status_message_handler.last_message == "first message");
+		auto message_lifetime2 = status_line.show_message_until_finished("second message");
+		REQUIRE(status_message_handler.last_message == "second message");
+		auto message_lifetime3 = status_line.show_message_until_finished("third message");
+		REQUIRE(status_message_handler.last_message == "third message");
+
+		WHEN("the latest message goes out of scope") {
+			message_lifetime3.reset();
+
+			THEN("the most recent active status message is restored") {
+				REQUIRE(status_message_handler.last_message == "second message");
+
+				THEN("a further destruction causes the then latest message to be restored") {
+					message_lifetime2.reset();
+					REQUIRE(status_message_handler.last_message == "first message");
```

Comment:
I'd add one final check to see what happens when the last remaining lifetime ends.